packages/back-end/src/events/handlers/slack/slack-event-handler-utils.ts

```diff
@@ -16,71 +16,74 @@ import { APP_ORIGIN } from "../../../util/secrets";
 
 // region Filtering
 
-/**
- * Gets all current and previous tags for the event
- * @param event
- */
-export const getTagsForNotificationEvent = (
-  event: NotificationEvent
-): string[] => {
-  switch (event.event) {
-    case "feature.created":
-      return event.data.current.tags || [];
-
-    case "feature.updated":
-      return uniq(
-        (event.data.current.tags || []).concat(event.data.previous.tags || [])
-      );
-
-    case "feature.deleted":
-      return event.data.previous.tags || [];
-  }
-};
-
-/**
- * Gets current and previous projects
- * @param event
- */
-export const getProjectsForNotificationEvent = (
-  event: NotificationEvent
-): string[] => {
-  switch (event.event) {
-    case "feature.created":
-      return event.data.current.project ? [event.data.current.project] : [];
-
-    case "feature.updated":
-      return uniq(
-        (event.data.current.project ? [event.data.current.project] : []).concat(
-          event.data.previous.project ? [event.data.previous.project] : []
-        )
-      );
-
-    case "feature.deleted":
-      return event.data.previous.project ? [event.data.previous.project] : [];
-  }
+type DataForNotificationEvent = {
+  filterData: {
+    tags: string[];
+    projects: string[];
+    environments: string[];
+  };
+  slackMessage: SlackMessage;
 };
 
-/**
- * In the case of some resources, e.g. features, we want to do our own filtering for environments.
- * We need to include all environments in that case.
- * @param event
- */
-export const getEnvironmentsForNotificationEvent = (
-  event: NotificationEvent
-): string[] => {
+export const getDataForNotificationEvent = (
+  event: NotificationEvent,
+  eventId: string
+): DataForNotificationEvent | null => {
   switch (event.event) {
     case "feature.created":
-      return Object.keys(event.data.current.environmentSettings);
+      return {
+        filterData: {
+          tags: event.data.current.tags || [],
```

Comment:
What you have here is readable and understandable at a glance. However, I wonder if it's a bit verbose. For similar events such as the feature.* ones, you could consider a single payload builder. Something like:
```
  tags: uniq([
    ...(event.data?.current?.tags || []),
    ...(event.data?.previous?.tags || []),
  ]),
},
slackMessage: buildSlackMessageForFeatureEvent(event, eventId)
```

Don't need to do this, just an idea to reduce boilerplate in exchange for making it a bit trickier to read.

Reply:
Do you mean where `event` can be any kind of event, e.g. created, updated, or deleted, and trying to access a deleted event's `current` and a `created` event's `previous` as optionals (where these properties don't exist on those event types)? If that's the case, while your suggestion works in plain JavaScript, it would cause a type error in TypeScript. Here's an example error if you try to do `event.data.current` or `event?.data?.current`:

```
Property 'current' does not exist on type '{ current: FeatureInterface; } | { current: FeatureInterface; previous: FeatureInterface; } | { previous: FeatureInterface; } | { ...; } | { ...; } | { ...; }'.   Property 'current' does not exist on type '{ previous: FeatureInterface; }'.
```

For the data modelling, I chose to use union types instead of optionals or `any` so we have more confidence in the data, which results in a more verbose approach, so I wouldn't be able to group them for a single event with something like `tags: uniq([ ...(event.data?.current?.tags || []), ...(event.data?.previous?.tags || []) ])` as suggested.

Reply:
yep, that is what I meant (event can be "created", "updated", or "deleted", with shared logic for building the payload). If we're using explicit union types then my suggestion is less clean (relies on optionals or coercion). Either way you want to go here is fine with me, just presenting an idea to reduce boiler plate at the expense of cleanness.